### PR TITLE
[Silabs] Update ThreadStackManagerImpl

### PR DESCRIPTION
**Pull request by lpbeliveau-silabs**
Update of Silabs code to use latest Simplicity SDK changes.

There IS a small change to the `GenericThreadStackManagerImpl_OpenThread` where I added a helper to allow to call parts of DoInit from a ThreadStackManagerImpl without code dupplication.

### CODE REVIEW

File: examples/platform/silabs/BaseApplication.cpp [hunk 1]
```diff
@@ -750,15 +790,42 @@ void BaseApplication::ScheduleFactoryReset()
...
+        PlatformMgr().HandleServerShuttingDown(); // HandleServerShuttingDown calls OnShutdown() which is only implemented for the basic information cluster it seems. And triggers and Event flush, which is not relevant when there are no fabrics left
         ConfigurationMgr().InitiateFactoryReset();
     });
 }
 
+void BaseApplication::DoProvisioningReset()
+{
+    PlatformMgr().ScheduleWork([](intptr_t) {
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD_SRP_CLIENT
```
Comment (jmartinez-silabs) [inline examples/platform/silabs/BaseApplication.cpp:804]: Can we put this reset code into a method and both DoFactoryReset and this call?

It seems like the same code other than the device reset at the end for the DoFactoryReset

File: src/platform/silabs/efr32/ThreadStackManagerImpl_CMSIS.hpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+template <class ImplClass>
+void ThreadStackManagerImpl_CMSIS<ImplClass>::_UnlockThreadStack(void)
+{
+}
+
+#if CHIP_DEVICE_CONFIG_ENABLE_THREAD_SRP_CLIENT
+template <class ImplClass>
+void ThreadStackManagerImpl_CMSIS<ImplClass>::_WaitOnSrpClearAllComplete()
+{
+    // Only 1 task can be blocked on a srpClearAll request
```
Comment (jmartinez-silabs) [inline src/platform/silabs/efr32/ThreadStackManagerImpl_CMSIS.hpp:68]: Do we have to keep all the commented code in this file

File: third_party/silabs/efr32_sdk.gni [hunk 3]
```diff
@@ -475,6 +477,25 @@ template("efr32_sdk") {
       ]
     }
 
+    if (defined(invoker.chip_enable_openthread) &&
+        invoker.chip_enable_openthread) {
+      defines += [
+        "SL_OT_ENABLE=1",
+
+        #TODO: Remove these defines onse sl_openthread_rtos_config is included in sl_ot_rtos_adaptation.c
```
Comment (jmartinez-silabs) [inline third_party/silabs/efr32_sdk.gni:485]:
```suggestion
        #TODO: Remove these defines onse sl_openthread_rtos_config is included in sl_ot_rtos_adaptation.c
```
```suggestion
        #TODO: Remove these defines once sl_openthread_rtos_config is included in sl_ot_rtos_adaptation.c
```

**Review by jmartinez-silabs — APPROVED** [review]

File: src/platform/silabs/efr32/ThreadStackManagerImpl_CMSIS.hpp (hunk 2 above)
Comment (lpbeliveau-silabs) [reply src/platform/silabs/efr32/ThreadStackManagerImpl_CMSIS.hpp:68]: Actually, we no not need this file... must have slipped through

File: examples/platform/silabs/BaseApplication.h [hunk 4]
```diff
@@ -156,6 +158,18 @@ class BaseApplication
...
 #endif
 
+    /**
+     * @brief Updates the static boolean isCommissioned to the desired state
+     *
+     */
+    static void UpdateCommissioningStatus(bool newState);
+
+    /**
+     * @brief Called when the last Fabric is removed, clears all Fabric related data and Thread  Wifi provision.
```
Comment (rcasallas-silabs) [inline examples/platform/silabs/BaseApplication.h:168]:
```suggestion
     * @brief Called when the last Fabric is removed, clears all Fabric related data, including Thread and Wifi provision.
```

File: examples/platform/silabs/BaseApplication.h [hunk 5]
```diff
@@ -156,6 +158,18 @@ class BaseApplication
...
+     * @note This function preserves some NVM3 data that is not Fabric scoped like Attribute Value or Boot Count.
```
Comment (rcasallas-silabs) [inline examples/platform/silabs/BaseApplication.h:169]:
```suggestion
     * @note This function preserves some NVM3 data that is not Fabric scoped, like Attribute Value or Boot Count.
```

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 6]
```diff
@@ -1088,29 +1088,21 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_GetPollPeriod(u
     return CHIP_NO_ERROR;
 }
 
+/**
+ * @brief Helper that sets callbacks for OpenThread state changes and configures the Thread stack.
+ * Assigns mOTInst to and instance and configures the OT stack on a device by setting state change callbacks enabling features
```
Comment (rcasallas-silabs) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1093]:
```suggestion
 * Assigns mOTInst to an instance, and configures the OT stack on a device by setting state change callbacks enabling features
```

File: src/platform/silabs/efr32/ThreadStackManagerImpl.cpp [hunk 7]
```diff
@@ -35,73 +36,127 @@
...
+#include <mbedtls/platform.h>
+
+extern "C" {
+#include "platform-efr32.h"
+}
+
 // Forward declaration
 extern "C" otInstance * otGetInstance(void);
 
+#if CHIP_DEVICE_CONFIG_THREAD_ENABLE_CLI
```
Comment (rcasallas-silabs) [inline src/platform/silabs/efr32/ThreadStackManagerImpl.cpp:49]: Looks like you can group several extern "C" here, like:
```
extern "C" {
#include "platform-efr32.h"
...
otInstance * otGetInstance(void);
#if CHIP_DEVICE_CONFIG_THREAD_ENABLE_CLI
void otAppCliInit(otInstance * aInstance);
...
}
#if CHIP_DEVICE_CONFIG_THREAD_ENABLE_CLI
```

**Review by rcasallas-silabs — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Pull request merged**